File: Data-Structures/Arrays/Exercises/Valid-Anagram/Solution-1/solution1.py
import unittest

###### IMPLEMENTATION #######

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        checked = set()

        for c in s:
            if c not in checked:
                if s.count(c) != t.count(c):
                    return False
                    
                checked.add(c)
        
        return True

###### TESTING #######

class TestIsAnagram(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().isAnagram("anagram", "nagaram"), True)
    
    def test_2(self):
        self.assertEqual(Solution().isAnagram("rat", "car"), False)


if __name__ == '__main__':
    unittest.main()

File: Data-Structures/Arrays/Exercises/Valid-Anagram/Solution-1/solution1.cpp
#include <iostream>
#include <set>
#include <iostream>
#include <algorithm>
#include <string>

using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.length() != t.length())
            return false;

        set<char> checked;

        for (auto c : s) 
            if (checked.find(c) == checked.end()) {
                if (count(s.begin(), s.end(), c) != count(t.begin(), t.end(), c))
                    return false;

                checked.insert(c);
            }
        return true;
    }
};

int main() {
    cout << Solution().isAnagram("anagram", "nagaram") << "\n";
    cout << Solution().isAnagram("rat", "car") << "\n";

    return 0;
}